featurebyte/query_graph/sql/aggregator/non_tile_window.py

```diff
@@ -0,0 +1,359 @@
+"""
+SQL generation for non-tile window aggregation
+"""
+
+from __future__ import annotations
+
+from typing import Any, Dict, Optional, Tuple, cast
+
+from dataclasses import dataclass
+
+from sqlglot import expressions
+from sqlglot.expressions import Select, alias_, select
+
+from featurebyte.enum import InternalName, SourceType, SpecialColumnName
+from featurebyte.query_graph.model.feature_job_setting import FeatureJobSetting
+from featurebyte.query_graph.sql.adapter import get_sql_adapter
+from featurebyte.query_graph.sql.aggregator.base import (
+    AggregationResult,
+    LeftJoinableSubquery,
+    NonTileBasedAggregator,
+)
+from featurebyte.query_graph.sql.aggregator.range_join import (
+    LeftTable,
+    RightTable,
+    range_join_tables,
+)
+from featurebyte.query_graph.sql.ast.literal import make_literal_value
+from featurebyte.query_graph.sql.common import CteStatement, CteStatements, quoted_identifier
+from featurebyte.query_graph.sql.feature_job import get_previous_job_epoch_expr_from_settings
+from featurebyte.query_graph.sql.groupby_helper import GroupbyColumn, GroupbyKey, get_groupby_expr
+from featurebyte.query_graph.sql.specifications.non_tile_window_aggregate import (
+    NonTileWindowAggregateSpec,
+)
+
+WindowType = int
+OffsetType = Optional[int]
+FeatureJobSettingsType = FeatureJobSetting
+EntityIDsType = Tuple[str, ...]
+RequestTableKeyType = Tuple[WindowType, OffsetType, FeatureJobSettingsType, EntityIDsType]
+
+
+@dataclass
+class ProcessedRequestTable:
+    """
+    Processed request table
+    """
+
+    name: str
+    aggregation_spec: NonTileWindowAggregateSpec
+
+
+class NonTileRequestTablePlan:
+    """
+    Responsible for processing request table to have distinct serving names and window ranges
+    """
+
+    def __init__(self, source_type: SourceType) -> None:
+        self.processed_request_tables: Dict[RequestTableKeyType, ProcessedRequestTable] = {}
+        self.adapter = get_sql_adapter(source_type)
+
+    def add_aggregation_spec(self, aggregation_spec: NonTileWindowAggregateSpec) -> None:
+        """
+        Update plan to account for the aggregation spec
+
+        Parameters
+        ----------
+        aggregation_spec: NonTileWindowAggregateSpec
+            Aggregation spec
+        """
+        key = self._get_request_table_key(aggregation_spec)
+        if key not in self.processed_request_tables:
+            window_spec = f"W{aggregation_spec.window}"
+            if aggregation_spec.offset is not None:
+                window_spec += f"_O{aggregation_spec.offset}"
+            job_settings = aggregation_spec.parameters.feature_job_setting
```

Comment:
Even there is no tile for this new type of aggregate, do we still need the feature job setting?

Reply:
I think we still need it for the feature materialize job. In historical serving we use the same feature job setting to simulate the effect of scheduled feature jobs to ensure online / offline consistency.

Reply:
Ok. Now i got it. Thanks.